### PR TITLE
Continue Tail simulations

### PR DESCRIPTION
**Pull request by ch1bo**
* Store transaction confirmation times as CSV result on every `hydra-tail-simulation run`
* Add a `plot-txs.mjs` script to plot transaction confirmation times
* Add `100`, `1000` and `20000` slot-compression datasets from a more recent `blocks.json`
* Add `--pro-active-snapshot` to have clients perform snapshots at some window limit, e.g. `0.8` means snapshots are created when reaching 80% of the payment window
* Add `hydra-tail-simulation analyze` to re-analyze previous results CSVs
* *BREAKING* Dropped `Analyze` sections on throughput as this was requiring simulation summary (and not easy to store/add to the `analyze` command)

### CODE REVIEW

File: scripts/tail/plot.mjs [hunk 1]
```diff
@@ -1,11 +1,11 @@
-import { ChartJSNodeCanvas } from 'chartjs-node-canvas';
+import chartjs from 'chartjs-node-canvas';
 import fs from 'fs';
 import assert from 'assert';
 import { readCsvFileSync } from './utils.mjs';
 
 const width = 960;
 const height = 500;
-const chartJSNodeCanvas = new ChartJSNodeCanvas({ width, height });
+const chartJSNodeCanvas = new chartjs.ChartJSNodeCanvas({ width, height });
```
Comment (KtorZ) [inline scripts/tail/plot.mjs:8]: Hmmm. Okay :shrug:

File: src/Hydra/Tail/Simulation.hs [hunk 2]
```diff
@@ -1,326 +1,309 @@
...
+        client <$ connectClient client server
+    void $
+        async $
+            concurrently_
+                (runServer trServer opts server)
+                (forConcurrently_ clients (runClient trClient (trim events) serverId opts))
+    -- XXX(SN): This does not take into account that there might still be clients
+    -- running and this likely leads to a differing number of confirmed
+    -- transactions for the same events dataset with different parameters (e.g.
+    -- payment window or pro-active snapshot limit)
```
Comment (KtorZ) [inline src/Hydra/Tail/Simulation.hs:127]: Indeed :+1:. The initial version of the simulation code was waiting for all clients to terminate. Despite being quite long with some parameters, it was also quite wrong, for after a certain point (the end of the planned schedule), there's no more events emitted by clients. The server is merely catching up with the remaining clients, giving completely wrong results over the observation window.

File: src/Hydra/Tail/Simulation.hs [hunk 3]
```diff
@@ -1,326 +1,309 @@
...
+        }
+  where
+    numberOfConfirmedTransactions = length confirmationTimes
+
+    confirmationTimes =
+        mapMaybe (convertConfirmationTime . snd) . maybeDiscardEdges $ Map.toList txs
+
+    maybeDiscardEdges xs = case discardEdges of
+        Nothing -> xs
+        Just n -> filter (\(TxRef{slot},_) -> slot > n && slot < (length xs - n)) xs
```
Comment (KtorZ) [inline src/Hydra/Tail/Simulation.hs:218]: :+1:

File: src/Hydra/Tail/Simulation.hs [hunk 4]
```diff
@@ -1,326 +1,309 @@
...
-  -- ↓↓↓ Server messages ↓↓↓
-  --
+    percentConfirmedWithinTenthOfSlot =
+        length (filter (< 0.1) confirmationTimes) `percentOf` numberOfConfirmedTransactions
 
-  | NotifyTx !MockTx
-  -- ^ The server will notify concerned clients with transactions they have subscribed to.
-  -- How clients subscribe and how the server is keeping track of the subscription is currently
-  -- out of scope and will be explored at a later stage.
+    percentOf a b = (fromIntegral a :: Double) / fromIntegral b
```
Comment (KtorZ) [inline src/Hydra/Tail/Simulation.hs:238]: :+1:

File: src/Hydra/Tail/Simulation.hs [hunk 5]
```diff
@@ -329,193 +312,226 @@ data TraceTailSimulation
...
+            if not (null blocked)
+                then do
+                    withTMVar_ registry $
+                        execStateT $ do
+                            forM_ (zip [0 ..] (Map.elems blocked)) $ \case
+                                (ix, (client, Just NeedSnapshot)) -> do
+                                    lift $ sendTo multiplexer client NeedSnapshot
+                                    modifyM $
+                                        updateF client $ \(_st, balance, mailbox, pending) -> do
+                                            -- NOTE(SN): This seems to be only re-enqueuing it for the sending client!?
```
Comment (KtorZ) [inline src/Hydra/Tail/Simulation.hs:392]: Not necessarily the sending client, but _one (and only one) of the participants_. Indeed, the transaction is marked as pending if one of the participants (sender or recipients) are blocked (require snapshots). From there, we can make two observations:

- For the transaction to succeed, both clients need to have performed their snapshot. 
- There's no "ledger" on the server, so if we re-enqueue a transaction per blocked participant, we'll may replay the transaction more than once! 

Thus, it is sufficient (and necessary) to re-enqueue the transaction only once, for one of the participant. Once that participant is done snapshotting, it'll retry the transaction and from here we really have two scenarios:

a) That client happened to be the last one in the transaction which had to snapshot, and the transaction can now proceed.

b) There's another client which is still blocked. Then, the transaction will be re-enqueued again, for that client (or at least, for the first other block client). 

Eventually, once all clients are done, it goes through.

File: src/Hydra/Tail/Simulation.hs [hunk 6]
```diff
@@ -329,193 +312,226 @@ data TraceTailSimulation
...
+                                    let newBalance = modifyCurrent (\x -> x - sent tx) balance
+                                    -- Check whether we should send 'NeedSnapshot' for pro-active
+                                    -- snapshotting (after the tx). Logically this would be done on the
+                                    -- client-side but we have the payment window 'registry' only on the
+                                    -- server for now.
+                                    if inProactiveSnapshotLimit options newBalance
+                                        then do
+                                            sendTo multiplexer clientId NeedSnapshot
+                                            pure $ Just (Blocked, newBalance, mailbox, pending)
+                                        else pure $ Just (st, newBalance, mailbox, pending)
```
Comment (KtorZ) [inline src/Hydra/Tail/Simulation.hs:424]: Ok :+1:

File: src/Hydra/Tail/Simulation.hs [hunk 7]
```diff
@@ -329,193 +312,226 @@ data TraceTailSimulation
...
+  where
+    absBalance = abs $ current - initial
+
+    limit w frac = fromDouble (toDouble (lovelace w) * frac)
+
+    toDouble :: Lovelace -> Double
+    toDouble = fromInteger . unLovelace
+
+    fromDouble :: Double -> Lovelace
+    fromDouble = Lovelace . truncate
```
Comment (KtorZ) [inline src/Hydra/Tail/Simulation.hs:519]: :+1:

File: src/Hydra/Tail/Simulation.hs [hunk 8]
```diff
@@ -525,145 +541,165 @@ instance Exception UnknownClient
...
+        fmap Ada $
+            state $
+                frequency
+                    [ (122, randomR (1, 10))
+                    , (144, randomR (10, 100))
+                    , (143, randomR (100, 1000))
+                    , (92, randomR (1000, 10000))
+                    , (41, randomR (10000, 100000))
+                    , (12, randomR (100000, 1000000))
+                    ]
```
Comment (KtorZ) [inline src/Hydra/Tail/Simulation.hs:667]: Ah! I thought I had pushed updated weights already. Those are old now.

**Review by KtorZ — APPROVED** [review]

File: scripts/tail/plot.mjs (hunk 1 above)
Comment (ch1bo) [reply scripts/tail/plot.mjs:8]: Not sure why, but my nodejs 14 was having problems with this module and suggested to do it that way.

File: src/Hydra/Tail/Simulation.hs (hunk 5 above)
Comment (ch1bo) [reply src/Hydra/Tail/Simulation.hs:392]: Thanks for the explanation. I'll copy it into the code before merging this as it explains the `ix == 0`

File: src/Hydra/Tail/Simulation.hs (hunk 8 above)
Comment (ch1bo) [reply src/Hydra/Tail/Simulation.hs:667]: Good to know. Haven't used the synthetic `prepare` in any of my experiments though.

**Pull request merged**